File: widgets/lib-arch/src/main/java/com/yyxnb/lib_arch/delegate/FragmentDelegate.java
package com.yyxnb.lib_arch.delegate;

import android.annotation.SuppressLint;
import android.arch.lifecycle.Lifecycle;
import android.arch.lifecycle.OnLifecycleEvent;
import android.content.Context;
import android.content.res.Configuration;
import android.graphics.Color;
import android.os.Build;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentActivity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.Window;

import com.yyxnb.lib_arch.annotations.BarStyle;
import com.yyxnb.lib_arch.annotations.BindRes;
import com.yyxnb.lib_arch.base.IActivity;
import com.yyxnb.lib_arch.base.IFragment;
import com.yyxnb.lib_arch.config.ArchConfig;
import com.yyxnb.lib_arch.config.ArchManager;
import com.yyxnb.lib_arch.helper.BusHelper;
import com.yyxnb.lib_arch.bean.MsgEvent;
import com.yyxnb.lib_arch.constants.ArgumentKeys;
import com.yyxnb.lib_common.action.HandlerAction;
import com.yyxnb.lib_common.interfaces.ILifecycle;
import com.yyxnb.util_core.StatusBarUtils;

import java.util.Objects;

/**
 * ================================================
 * 作    者：yyx
 * 版    本：1.0
 * 日    期：2020/11/21
 * 历    史：
 * 描    述：Fragment 代理
 * ================================================
 */
public class FragmentDelegate implements ILifecycle, HandlerAction {

    public FragmentDelegate(IFragment iFragment) {
        this.iFragment = iFragment;
        this.mFragment = (Fragment) iFragment;
        mLazyDelegate = new FragmentLazyDelegate(mFragment);
    }

    private IFragment iFragment;
    private IActivity iActivity;

    private View mRootView;

    private Fragment mFragment;
    private FragmentActivity mActivity;

    private FragmentLazyDelegate mLazyDelegate;

    private static final ArchConfig config = ArchManager.getInstance().getConfig();
    private int layoutRes = 0;
    private boolean statusBarTranslucent = config.isStatusBarTranslucent();
    private boolean fitsSystemWindows = config.isFitsSystemWindows();
    private int statusBarColor = config.getStatusBarColor();
    private int statusBarDarkTheme = config.getStatusBarStyle();
    private int swipeBack = config.getSwipeBack();
    private boolean subPage;
    private boolean needLogin = config.isNeedLogin();

    public FragmentActivity getActivity() {
        return mActivity;
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_CREATE)
    public void onAttach(Context context) {
        mActivity = (FragmentActivity) context;
        if (!(mActivity instanceof IActivity)) {
            throw new IllegalArgumentException("Activity请实现IActivity接口");
        }
        iActivity = (IActivity) mActivity;
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_CREATE)
    public void onCreate(Bundle savedInstanceState) {
        mLazyDelegate.onCreate(savedInstanceState);
        initAttributes();
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_CREATE)
    @SuppressLint("ClickableViewAccessibility")
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        if (null == mRootView) {
            if (layoutRes != 0 || iFragment.initLayoutResId() != 0) {
                mRootView = inflater.inflate(layoutRes == 0 ? iFragment.initLayoutResId()
                        : layoutRes, container, false);
            }
        } else {
            //  二次加载删除上一个子view
            ViewGroup viewGroup = (ViewGroup) mRootView;
            viewGroup.removeView(mRootView);
        }
        mRootView.setOnTouchListener((v, event) -> {
            mActivity.onTouchEvent(event);
            return false;
        });
        return mRootView;
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_CREATE)
    public void onActivityCreated(Bundle savedInstanceState) {
        mLazyDelegate.onActivityCreated(savedInstanceState);
        setNeedsStatusBarAppearanceUpdate();
    }

    /**
     * viewpager调用 {@link BindRes} {@link FragmentDelegate#subPage 为 true}
     */
    public void setUserVisibleHint(boolean isVisibleToUser) {
        mLazyDelegate.setUserVisibleHint(isVisibleToUser);
    }

    /**
     * show/hide调用
     */
    public void onHiddenChanged(boolean hidden) {
        mLazyDelegate.onHiddenChanged(hidden);
    }

    /**
     * 屏幕方向发生改变时
     */
    public void onConfigurationChanged(Configuration newConfig) {
        mLazyDelegate.onConfigurationChanged(newConfig);
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_RESUME)
    public void onResume() {
        mLazyDelegate.onResume();
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_PAUSE)
    public void onPause() {
        mLazyDelegate.onPause();
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_DESTROY)
    public void onDestroy() {
        removeCallbacks();
        mLazyDelegate.onDestroy();
        iFragment = null;
        iActivity = null;
        mActivity = null;
        mFragment = null;
        mRootView = null;
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_DESTROY)
    public void onDestroyView() {
    }

    /**
     * 加载注解设置
     */
    public void initAttributes() {
        post(() -> {
            final BindRes bindRes = iFragment.getClass().getAnnotation(BindRes.class);
            if (bindRes != null) {
                layoutRes = bindRes.layoutRes();
                fitsSystemWindows = bindRes.fitsSystemWindows();
                statusBarTranslucent = bindRes.statusBarTranslucent();
                swipeBack = bindRes.swipeBack();
                subPage = bindRes.subPage();
                if (bindRes.statusBarStyle() != BarStyle.NONE) {
                    statusBarDarkTheme = bindRes.statusBarStyle();
                }
                if (bindRes.statusBarColor() != 0) {
                    statusBarColor = bindRes.statusBarColor();
                }
                needLogin = bindRes.needLogin();
                // 如果需要登录，并且处于未登录状态下，发送通知
                if (needLogin) {
                    BusHelper.post(new MsgEvent(ArgumentKeys.NEED_LOGIN_CODE));
                }
            }
        });
    }

    /**
     * 更新状态栏样式
     */
    public void setNeedsStatusBarAppearanceUpdate() {

        // 子页面不作做处理
        if (subPage) {
            return;
        }
        // 侧滑返回
        iActivity.setSwipeBack(swipeBack);

        // 文字颜色
        int statusBarStyle = statusBarDarkTheme;
        StatusBarUtils.setStatusBarStyle(getWindow(), statusBarStyle == BarStyle.DARK_CONTENT);

        // 隐藏 or 不留空间 则透明
        if (!fitsSystemWindows) {
            StatusBarUtils.setStatusBarColor(getWindow(), Color.TRANSPARENT);
        } else {
            int statusBarColor = this.statusBarColor;

            //不为深色
            boolean shouldAdjustForWhiteStatusBar = !StatusBarUtils.isBlackColor(statusBarColor, 176);

            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                shouldAdjustForWhiteStatusBar = shouldAdjustForWhiteStatusBar && statusBarStyle == BarStyle.LIGHT_CONTENT;
            }
            // 如果状态栏处于白色且状态栏文字也处于白色，避免看不见
            if (shouldAdjustForWhiteStatusBar) {
                statusBarColor = config.getShouldAdjustForWhiteStatusBar();
            }

            StatusBarUtils.setStatusBarColor(getWindow(), statusBarColor);
        }
        StatusBarUtils.setStatusBarTranslucent(getWindow(), statusBarTranslucent, fitsSystemWindows);
    }

    public Window getWindow() {
        return mActivity.getWindow();
    }

    public Bundle initArguments() {
        Bundle args = mFragment.getArguments();
        if (args == null) {
            args = new Bundle();
            mFragment.setArguments(args);
        }
        return args;
    }

    public void finish() {
        mActivity.onBackPressed();
    }

    public FragmentLazyDelegate getLazyDelegate() {
        return mLazyDelegate;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        FragmentDelegate that = (FragmentDelegate) o;
        return iFragment.equals(that.iFragment);
    }

    @Override
    public int hashCode() {
        return Objects.hash(iFragment);
    }

}

File: lib_arch/src/main/java/com/yyxnb/arch/delegate/FragmentDelegate.kt
package com.yyxnb.arch.delegate

import android.annotation.SuppressLint
import android.content.Context
import android.content.res.Configuration
import android.graphics.Color
import android.os.Build
import android.os.Bundle
import android.view.*
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentActivity
import androidx.lifecycle.DefaultLifecycleObserver
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.OnLifecycleEvent
import com.yyxnb.arch.annotations.BarStyle
import com.yyxnb.arch.annotations.BindRes
import com.yyxnb.arch.annotations.SwipeStyle
import com.yyxnb.arch.base.IActivity
import com.yyxnb.arch.base.IFragment
import com.yyxnb.arch.common.ArchConfig
import com.yyxnb.arch.common.Bus.post
import com.yyxnb.arch.common.MsgEvent
import com.yyxnb.common.utils.MainThreadUtils.post
import com.yyxnb.common.utils.StatusBarUtils.isBlackColor
import com.yyxnb.common.utils.StatusBarUtils.setStatusBarColor
import com.yyxnb.common.utils.StatusBarUtils.setStatusBarStyle
import com.yyxnb.common.utils.StatusBarUtils.setStatusBarTranslucent
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.MainScope
import kotlinx.coroutines.NonCancellable.cancel
import kotlinx.coroutines.cancel
import kotlinx.coroutines.isActive
import java.util.*

/**
 * Fragment 代理
 *
 * @author yyx
 */
class FragmentDelegate(private var iFragment: IFragment?) : DefaultLifecycleObserver, CoroutineScope by MainScope() {

    private var iActivity: IActivity? = null
    private var mRootView: View? = null
    private var mFragment: Fragment?
    private var mActivity: FragmentActivity? = null

    private val mLazyDelegate: FragmentLazyDelegate
    private var layoutRes = 0
    private var statusBarTranslucent = ArchConfig.statusBarTranslucent
    private var fitsSystemWindows = ArchConfig.fitsSystemWindows
    private var statusBarColor = ArchConfig.statusBarColor
    private var statusBarDarkTheme = ArchConfig.statusBarStyle
    private var swipeBack = SwipeStyle.EDGE
    private var subPage = false
    private var needLogin = false

    init {
        mFragment = iFragment as Fragment?
        mLazyDelegate = FragmentLazyDelegate(mFragment)
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_CREATE)
    fun onAttach(context: Context?) {
        mActivity = context as FragmentActivity?
        require(mActivity is IActivity) { "Activity请实现IActivity接口" }
        iActivity = mActivity as IActivity?
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_CREATE)
    fun onCreate(savedInstanceState: Bundle?) {
        mLazyDelegate.onCreate(savedInstanceState)
        initAttributes()
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_CREATE)
    @SuppressLint("ClickableViewAccessibility")
    fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        if (null == mRootView) {
            if (layoutRes != 0 || iFragment!!.initLayoutResId() != 0) {
                mRootView = inflater.inflate(if (layoutRes == 0) iFragment!!.initLayoutResId() else layoutRes, container, false)
            }
        } else {
            //  二次加载删除上一个子view
            val viewGroup = mRootView as ViewGroup
            viewGroup.removeView(mRootView)
        }
        mRootView!!.setOnTouchListener { v: View?, event: MotionEvent? ->
            mActivity!!.onTouchEvent(event)
            false
        }
        return mRootView
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_CREATE)
    fun onActivityCreated(savedInstanceState: Bundle?) {
        mLazyDelegate.onActivityCreated(savedInstanceState)
        setNeedsStatusBarAppearanceUpdate()
    }

    /**
     * viewpager调用 [BindRes] [为 true][FragmentDelegate.subPage]
     */
    fun setUserVisibleHint(isVisibleToUser: Boolean) {
        mLazyDelegate.setUserVisibleHint(isVisibleToUser)
    }

    /**
     * show/hide调用
     */
    fun onHiddenChanged(hidden: Boolean) {
        mLazyDelegate.onHiddenChanged(hidden)
    }

    /**
     * 屏幕方向发生改变时
     */
    fun onConfigurationChanged(newConfig: Configuration?) {
        mLazyDelegate.onConfigurationChanged(newConfig)
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_RESUME)
    fun onResume() {
        mLazyDelegate.onResume()
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_PAUSE)
    fun onPause() {
        mLazyDelegate.onPause()
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_DESTROY)
    fun onDestroy() {
        mLazyDelegate.onDestroy()
        iFragment = null
        iActivity = null
        mActivity = null
        mFragment = null
        mRootView = null
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_DESTROY)
    fun onDestroyView() {
        // 取消协程
        if (isActive) {
            cancel()
        }
    }

    /**
     * 加载注解设置
     */
    fun initAttributes() {
        post(Runnable {
            val bindRes = iFragment?.javaClass!!.getAnnotation(BindRes::class.java)
            if (bindRes != null) {
                layoutRes = bindRes.layoutRes
                fitsSystemWindows = bindRes.fitsSystemWindows
                statusBarTranslucent = bindRes.statusBarTranslucent
                swipeBack = bindRes.swipeBack
                subPage = bindRes.subPage
                if (bindRes.statusBarStyle != BarStyle.NONE) {
                    statusBarDarkTheme = bindRes.statusBarStyle
                }
                if (bindRes.statusBarColor != 0) {
                    statusBarColor = bindRes.statusBarColor
                }
                needLogin = bindRes.needLogin
                // 如果需要登录，并且处于未登录状态下，发送通知
                if (needLogin && !ArchConfig.needLogin) {
                    post(MsgEvent(ArchConfig.NEED_LOGIN_CODE))
                }
            }
        })
    }

    /**
     * 更新状态栏样式
     */
    fun setNeedsStatusBarAppearanceUpdate() {

        // 子页面不作做处理
        if (subPage) {
            return
        }
        // 侧滑返回
        iActivity!!.setSwipeBack(swipeBack)

        // 文字颜色
        val statusBarStyle = statusBarDarkTheme
        setStatusBarStyle(window, statusBarStyle == BarStyle.DARK_CONTENT)

        // 隐藏 or 不留空间 则透明
        if (!fitsSystemWindows) {
            setStatusBarColor(window, Color.TRANSPARENT)
        } else {
            var statusBarColor = statusBarColor

            //不为深色
            var shouldAdjustForWhiteStatusBar = !isBlackColor(statusBarColor, 176)
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                shouldAdjustForWhiteStatusBar = shouldAdjustForWhiteStatusBar && statusBarStyle == BarStyle.LIGHT_CONTENT
            }
            // 如果状态栏处于白色且状态栏文字也处于白色，避免看不见
            if (shouldAdjustForWhiteStatusBar) {
                statusBarColor = ArchConfig.shouldAdjustForWhiteStatusBar
            }
            setStatusBarColor(window, statusBarColor)
        }
        setStatusBarTranslucent(window, statusBarTranslucent, fitsSystemWindows)
    }

    val window: Window
        get() = mActivity!!.window

    fun initArguments(): Bundle {
        var args = mFragment!!.arguments
        if (args == null) {
            args = Bundle()
            mFragment!!.arguments = args
        }
        return args
    }

    fun finish() {
        mActivity!!.onBackPressed()
    }

    fun getLazyDelegate(): FragmentLazyDelegate = mLazyDelegate

    override fun equals(o: Any?): Boolean {
        if (this === o) {
            return true
        }
        if (o == null || javaClass != o.javaClass) {
            return false
        }
        val that = o as FragmentDelegate
        return iFragment == that.iFragment
    }

    override fun hashCode(): Int {
        return Objects.hash(iFragment)
    }

}